optuna/integration/tensorflow.py

```diff
@@ -0,0 +1,107 @@
+from __future__ import absolute_import
+
+import optuna
+
+
+try:
+    import tensorflow as tf
+    from tensorflow.train import SessionRunHook
+    _available = True
+except ImportError as e:
+    _import_error = e
+    # TensorFlowPruningHook is disabled because TensorFlow is not available.
+    _available = False
+    SessionRunHook = object
+
+
+class TensorFlowPruningHook(SessionRunHook):
+    """TensorFlow SessionRunHook to prune unpromising trials.
+
+    Example:
+
+        Add a pruning SessionRunHook for a TensorFlow's Estimator.
+
+        .. code::
+
+                pruning_hook = TensorFlowPruningHook(
+                    trial=trial,
+                    estimator=clf,
+                    metric="accuracy",
+                    is_higher_better=True,
+                    run_every_steps=10,
+                )
+                hooks = [pruning_hook]
+                tf.estimator.train_and_evaluate(
+                    clf,
+                    tf.estimator.TrainSpec(input_fn=train_input_fn, max_steps=500, hooks=hooks),
+                    eval_spec
+                )
+    Args:
+        trial:
+            A :class:`~optuna.trial.Trial` corresponding to the current evaluation of
+            the objective function.
+        estimator:
+            An estimator which you will use.
+        metric:
+            An evaluation metric for pruning, e.g., ``accuracy`` and ``loss``.
+        is_higher_better:
+           It should be True if you use a metric to be maximize such as ``loss``.
+        run_every_steps:
+           An interval to watch the summary file.
+    """
+    def __init__(self, trial, estimator, metric, is_higher_better, run_every_steps):
+        # type: (optuna.trial.Trial, tf.estimator.Estimator, str, bool, int) -> None
+
+        self.trial = trial
+        self.estimator = estimator
+        self.current_summary_step = -1
+        self.metric = metric
+        self.is_higher_better = is_higher_better
+        self._global_step_tensor = None
+        self._timer = tf.train.SecondOrStepTimer(every_secs=None, every_steps=run_every_steps)
+
+    def begin(self):
+        # type: () -> None
+
+        self._global_step_tensor = tf.train.get_global_step()
+
+    def before_run(self, run_context):
+        # type: (tf.train.SessionRunContext) -> tf.train.SessionRunArgs
+
+        del run_context
+        return tf.train.SessionRunArgs(self._global_step_tensor)
+
+    def after_run(self, run_context, run_values):
+        # type: (tf.train.SessionRunContext, tf.train.SessionRunValues) -> None
+
+        global_step = run_values.results
+        # Get eval metrics every n steps
+        if self._timer.should_trigger_for_step(global_step):
```

Comment:
last triggered steps were updated in `self._timer.should_trigger_for_step` automatically, but they weren't as far as I confirmed [the source code](https://github.com/tensorflow/tensorflow/blob/r1.12/tensorflow/python/training/basic_session_run_hooks.py#L118). 
Could you check this issue?

Reply:
Thanks, you are right. I implemented it referring to the [code](https://github.com/tensorflow/tensorflow/blob/b9e6bbc95bcffa481d29e31b448a03a91ba17eac/tensorflow/contrib/estimator/python/estimator/early_stopping.py#L445-L446), but forgot to run `self._timer.update_last_triggered_step(global_step)`.
I will add a line, and commit it after passing the tests.

Reply:
I fixed it!